File: src/deckgl_marimo/widget.js
import maplibregl from "https://esm.sh/maplibre-gl@4.7.1";

// --- deck.gl via standalone bundle (avoids ESM packaging issues) ---
const DECKGL_VERSION = "9.1.8";
const deckReady = (() => {
  if (window.deck) return Promise.resolve(window.deck);
  return new Promise((resolve, reject) => {
    const existing = document.querySelector(
      'script[src*="deck.gl/dist.min.js"]'
    );
    if (existing) {
      existing.addEventListener("load", () => resolve(window.deck));
      return;
    }
    const script = document.createElement("script");
    script.src = `https://unpkg.com/deck.gl@${DECKGL_VERSION}/dist.min.js`;
    script.onload = () => resolve(window.deck);
    script.onerror = () => reject(new Error("Failed to load deck.gl"));
    document.head.appendChild(script);
  });
})();

// Traitlet names that map to HexagonLayer props
const LAYER_PROPS = [
  "radius",
  "elevation_scale",
  "color_range",
  "extruded",
  "coverage",
  "upper_percentile",
  "pickable",
];

function getLayerProps(model) {
  return {
    radius: model.get("radius"),
    elevationScale: model.get("elevation_scale"),
    colorRange: model.get("color_range"),
    extruded: model.get("extruded"),
    coverage: model.get("coverage"),
    upperPercentile: model.get("upper_percentile"),
    pickable: model.get("pickable"),
  };
}

function buildLayer(deck, model) {
  const positions = model.get("positions") || [];
  return new deck.HexagonLayer({
    id: "hexagon-layer",
    data: positions,
    getPosition: (d) => d,
    ...getLayerProps(model),
  });
}

async function render({ model, el }) {
  const deck = await deckReady;

  // Container
  const container = document.createElement("div");
  container.style.width = "100%";
  container.style.height = model.get("map_height");
  container.classList.add("deckgl-marimo-container");
  el.appendChild(container);

  // MapLibre map
  const map = new maplibregl.Map({
    container,
    style: model.get("style_url"),
    center: [model.get("center_lon"), model.get("center_lat")],
    zoom: model.get("zoom"),
    pitch: model.get("pitch"),
    bearing: model.get("bearing"),
    antialias: true,
  });

  map.addControl(new maplibregl.NavigationControl(), "top-right");

  // deck.gl overlay
  const overlay = new deck.MapboxOverlay({
    layers: [buildLayer(deck, model)],
  });
  map.addControl(overlay);

  // React to traitlet changes — rebuild layer
  const layerTraitlets = ["positions", ...LAYER_PROPS];
  for (const name of layerTraitlets) {
    model.on(`change:${name}`, () => {
      overlay.setProps({ layers: [buildLayer(deck, model)] });
    });
  }

  // React to map_height changes
  model.on("change:map_height", () => {
    container.style.height = model.get("map_height");
    map.resize();
  });

  // Write viewport back to Python on moveend
  map.on("moveend", () => {
    const center = map.getCenter();
    model.set("viewport", {
      longitude: center.lng,
      latitude: center.lat,
      zoom: map.getZoom(),
      pitch: map.getPitch(),
      bearing: map.getBearing(),
    });
    model.save_changes();
  });

  // Cleanup
  return () => {
    overlay.finalize();
    map.remove();
  };
}

export default { render };

File: src/deckgl_marimo/widget.py
"""DeckGLHexagonWidget — anywidget wrapping MapLibre GL JS + deck.gl HexagonLayer."""

from __future__ import annotations

import pathlib
from typing import Any

import anywidget
import traitlets

from deckgl_marimo._data import dataframe_to_positions

DEFAULT_COLOR_RANGE = [
    [1, 152, 189],
    [73, 227, 206],
    [216, 254, 181],
    [254, 237, 177],
    [254, 173, 84],
    [209, 55, 78],
]

_HERE = pathlib.Path(__file__).parent


class DeckGLHexagonWidget(anywidget.AnyWidget):
    """A deck.gl HexagonLayer rendered on a MapLibre GL JS base map.

    Parameters
    ----------
    style_url : str
        MapLibre style URL (required, no default).
    data : DataFrame or list[dict], optional
        Source data with lat/lon columns.
    lat_col, lon_col : str
        Column names for latitude and longitude.
    center_lon, center_lat : float
        Initial map center.
    zoom, pitch, bearing : float
        Initial view state.
    radius, elevation_scale : int
        HexagonLayer parameters.
    color_range : list
        6-step color ramp as list of [r, g, b] lists.
    extruded, pickable : bool
        HexagonLayer flags.
    coverage, upper_percentile : float
        HexagonLayer parameters.
    map_height : str
        CSS height of the map container.
    """

    _esm = _HERE / "widget.js"
    _css = _HERE / "widget.css"

    # Map style
    style_url = traitlets.Unicode().tag(sync=True)

    # Positions data — [[lon, lat], ...]
    positions = traitlets.List(traitlets.List()).tag(sync=True)

    # View state
    center_lon = traitlets.Float(-1.4157).tag(sync=True)
    center_lat = traitlets.Float(52.2324).tag(sync=True)
    zoom = traitlets.Float(6.0).tag(sync=True)
    pitch = traitlets.Float(40.5).tag(sync=True)
    bearing = traitlets.Float(0.0).tag(sync=True)

    # HexagonLayer params
    radius = traitlets.Int(1000).tag(sync=True)
    elevation_scale = traitlets.Int(250).tag(sync=True)
    color_range = traitlets.List(DEFAULT_COLOR_RANGE).tag(sync=True)
    extruded = traitlets.Bool(True).tag(sync=True)
    coverage = traitlets.Float(1.0).tag(sync=True)
    upper_percentile = traitlets.Int(100).tag(sync=True)
    pickable = traitlets.Bool(True).tag(sync=True)

    # Layout
    map_height = traitlets.Unicode("600px").tag(sync=True)

    # Viewport read-back from JS (read-only on Python side)
    viewport = traitlets.Dict({}).tag(sync=True)

    def __init__(
        self,
        style_url: str,
        data: Any = None,
        lat_col: str = "lat",
        lon_col: str = "lon",
        **kwargs: Any,
    ) -> None:
        positions = []
        if data is not None:
            positions = dataframe_to_positions(data, lat_col=lat_col, lon_col=lon_col)
        super().__init__(style_url=style_url, positions=positions, **kwargs)
